contents/data/nosql/NoSQL.md

```diff
@@ -0,0 +1,437 @@
+    discountCode: "panda"
+  },
+  email: "panda@pandas.com",
+  ...
+}
+```
+
+In the case of more complicated
+<tooltip content="A type of relation where each item can be related to many other items, and these other items can also be related to many items on the other end">
+_many-to-many_
+</tooltip>
+relationships, relations are commonly stored using references, to avoid duplication of data.
+
+For example, items in an e-commerce website are related to the many customers through their carts.
+In these carts, it is much more space efficient to store references to the items, than the item documents themselves.
+
+In this example, the uniquely generated `_id` field for each item document in the items collection could be one such _reference_:
+```js
+{
+  type: "book",
+  price: 20,
+  popularity: 9.7,
+  _id: "9d1793bd491349n913847n93d"
+}
+```
+
+In the user's cart, we would simply store these `_id` _references_, which are used to lookup the item documents in the items collection later:
+```js
+cart: {
+  totalPrice: 100,
+  cartItems: [
+    "9d1793bd491349n913847n93d",
+    "9d1793bd491349n913847njh8",
+  ],
+  discountCode: "panda"
+}
+```
+
+
+
+<box type="info" light>
+For this reason, many NoSQL database solutions (e.g. MongoDB) implement a unique <code>id</code> field
+for each document by default.
+</box>
+
+---
+
+## Why NoSQL?
+
+<br>
+
+### 1. Highly Suited for Iterated Development
+Although less mature than SQL databases, NoSQL databases were designed to solve many of the emerging challenges in databases today.
+
+One of the most consequential impacts NoSQL has had was enabling faster iterated development.
+Given the highly flexible relational structure of NoSQL databases, and the schemaless format of documents 
+in NoSQL, this means that developers can adapt the database quicker to changing customer and business requirements.
+
+
+### 2. Easy Horizontal Scaling
+
+Another key benefit of NoSQL databases is the ability to scale horizontally (distributing workload across
+multiple servers), without discarding much of its key features.
+
+This is largely due to the schema-less architecture of such databases, allowing data to split across
+multiple servers easily and efficiently.
+
+For example, take the following collection of items with a `title`:
+
+```js
+[
+  {
+    title: "let's assume item titles are lexicographically distributed uniformly over the long run"
+    type: "book",
+    price: 20,
+    popularity: 9.7,
+  },
+  // item 2
+  // item 3
+  // ....
+  // item N
+]
+```
+
+Assuming we don't have relations from items to themselves inside these documents,
+we can split the collection like so:
+
+<img src="horizontalScalability.png" width="500" />
+
+As a result, the database access workload can be distributed evenly and efficiently across
+multiple servers easily.
+
+<box type="success" light>
+
+As businesses grow and see increasing amounts of web traffic, it is <em>crucial</em> that its databases can scale to meet
```

Comment:
I wonder why _web traffic_ was specifically mentioned here. It doesn't necessarily have to be web traffic, right?